[#66434] /ServiceProviderConfig endpoint responds with limited output without authorization.

ba1ash opened the pull request:
# Ticket
<!-- Provide the link to respective work package -->
https://community.openproject.org/work_packages/66434
<!-- Contributors: Please check our PR guide: https://www.openproject.org/docs/development/code-review-guidelines/#preparing-your-pull-request before opening a PR. -->

<!-- Reviewers: Please check our Review guide: https://www.openproject.org/docs/development/code-review-guidelines/#reviewing -->

# Merge checklist

- [x] Added/updated tests

ba1ash committed:
[#66434] /ServiceProviderConfig endpoint responds with limited output without authorization.

https://community.openproject.org/work_packages/66434

NobodysNightmare commented on app/controllers/scim_v2/scim_controller_mixins.rb:
```diff
@@ -30,29 +30,65 @@
-    def self.included(base)
-      base.prepend(Overwrites)
-    end
+    module ApplicationControllerMixin
```
:yellow_circle: Would it be possible that we split this file into two files? One file per mixin module?

The current layout is a bit surprising to me.

NobodysNightmare commented on app/controllers/scim_v2/scim_controller_mixins.rb:
```diff
@@ -30,29 +30,65 @@
+
+      module Overwrites
+        def show
+          if User.current == User.anonymous
```
:green_circle: Because I just saw a similar question being raised in one of our internal channels, what would you think about `User.current.anonymous?`

ba1ash replied:
The problem here is that `ServiceAccount#logged?` is defined as `false`. Therefore `User.current.anonymous?` responds with `true`... Should we consider changing `ServiceAccount#logged?` to be `true`?

NobodysNightmare replied:
oh... yeah, I guess that may have been an early error when implementing service accounts. Wondering where I took that default from, maybe from `SystemUser`?

Anyways, it sounds like we should change `ServiceAccount#logged?` to `true`, unless you already see that this would cause other harm that I am overlooking now.

ba1ash replied:
I will set `ServiceAccoung#logged?` to `true` in a separate PR.

NobodysNightmare commented on app/controllers/scim_v2/scim_controller_mixins.rb:
```diff
@@ -30,29 +30,65 @@
+      module Overwrites
+        def show
+          if User.current == User.anonymous
+            if warden.winning_strategy.blank? # it means authorization header was absent. So, there is no appropriate strategy
```
:yellow_circle: I am wondering whether this warden distinction matters at all.

If you are properly authenticated, we show you everything, if you are anonymous we show you the limited config.

Which case are we covering that would throw a `:warden` symbol here?

ba1ash replied:
The idea is to distinguish between two cases:
1. Authorization header is provided, but warden was not able to authenticate user with it. So, we respond with Unauthorized then. In this case there is a winning strategy.
2. Authorization header is not provided. Then there is no winning_strategy.

NobodysNightmare replied:
But then shouldn't this case be handled in the application controller already?

I.e. we render the 401, because authentication was attempted but unsuccessful, this sounds like a generic case that we always want to handle like that.

Then we could focus here in distinguishing between "truly" public requests and "truly" authenticated requests.

ba1ash replied:
You're right. It should be in application controller.

NobodysNightmare commented on lib/scimitar_schema_extension.rb:
```diff
@@ -48,6 +48,49 @@ class OpenProjectNameComplexType < Scimitar::ComplexTypes::Base
+    Scimitar::AuthenticationScheme.new(
+      type: "oauth2",
+      name: "OAuth2",
+      description: OpenProject::Static::Links.links[:sysadmin_docs][:scim_static_access_token_authentication_method][:href]
```
:yellow_circle: One more chance for me to advertise that there is a helper for these kinds of calls to the static links:

```suggestion
      description: OpenProject::Static::Links.url_for(:sysadmin_docs, :scim_static_access_token_authentication_method)
```

The main advantage being that this just semantically says "I want that link", without revealing the underlying data structure. For example there was already experiments going on, where this method would then return different links based on the current user's locale.

NobodysNightmare commented on lib/scimitar_schema_extension.rb:
```diff
@@ -48,6 +48,49 @@ class OpenProjectNameComplexType < Scimitar::ComplexTypes::Base
+    )
+  ].freeze
+
+  class LimitedServiceProviderConfiguration
```
:yellow_circle: I guess I'd also prefer more files for all these classes in this file. I noticed how I was irritated when I saw the `super` call below and didn't quite know where to scroll to find what "super" means.

NobodysNightmare commented on lib/scimitar_schema_extension.rb:
```diff
@@ -48,6 +48,49 @@ class OpenProjectNameComplexType < Scimitar::ComplexTypes::Base
+          version: "1"
+        ),
+        schemas: ["urn:ietf:params:scim:schemas:core:2.0:ServiceProviderConfig"],
+        authenticationSchemes: AUTHENTICATION_SCHEMES
```
I am now wondering whether the non-shortened version of the service provider configuration will have these authentication schemes available.

Probably not, because we don't seem to override that, correct?

ba1ash replied:
No, it is fine. We pass AUTHENTICATION_SCHEMES to ServiceProviderConfiguration as well in scimitar initializer.

NobodysNightmare commented on spec/requests/scim_v2/service_provider_config_spec.rb:
```diff
@@ -42,21 +42,60 @@
+                                         "sort" => { "supported" => false })
+      end
+
+      context '' do
```
:red_circle: Was it a forgotten TODO to name this context?

ba1ash replied:
Yes...

NobodysNightmare reviewed (changes requested):
I left some questions. Also there is quite a huge amount of Rubocop offenses in spec/requests/scim_v2/service_provider_config_spec.rb which seem fixable, even considering that the large object dumps we are using for expectations might fail _some_ other cops.

ba1ash committed:
[#66434] Update scimitar.

- Use custom_authenticator provied by scimitar to do authentication.
- Polish specs a bit.
- Rework OpenProject scimitar modifications structure.

Co-Authored-by: Jan Sandbrink <j.sandbrink@openproject.com>

NobodysNightmare commented on lib/open_project_scimitar/limited_service_provider_configuration.rb:
```diff
@@ -28,32 +28,29 @@
-        end
-      end
+    attr_accessor(
+      :authenticationSchemes,
```
I think rubocop complains about this not being called `authentication_schemes`. I am wondering: Is it required that the attribute accessor is camel cased?

ba1ash replied:
Eventually it must be camel cased in the returned json body.

NobodysNightmare replied:
I understand that. But we don't need to name our ruby methods after JSON structure, do we? :D

ba1ash replied:
We do not. Sure, let's change it.

ba1ash committed:
[#66434] Drop LimitedServiceProviderConfiguration.

No need in a class actually. Let's use a plain Hash instance instead.

NobodysNightmare commented on config/initializers/scimitar.rb:
```diff
@@ -31,17 +31,52 @@
+
+      warden = request.env["warden"]
+      user = warden.authenticate(scope: :scim_v2)
+      if user == nil
```
:green_circle: I am surprised that there is no cop for this:

```suggestion
      if user.nil?
```

NobodysNightmare reviewed (approved):
I found one minor thing while double reading the code, but LGTM otherwise.

ba1ash committed:
[#66434] Use .nil?.

Co-authored-by: Jan Sandbrink <j.sandbrink@openproject.com>
